### PR TITLE
Disable dev preview when dev command finished

### PR DESCRIPTION
**Pull request by alvaro-shopify**
### WHY are these changes introduced?

Extracted disable dev preview when dev command finishes

### WHAT is this pull request doing?
- Disable developer preview when `dev` commands finishes
-  Modify `ConcurrentOutput` component to finish the command (press `q` or `ctrl + c` using the `AbortController` instead of `treeKill`
- `dev` add a graceful finishing process to close every open external process. 
  - Include a timeout to force the finishing of the command
  - dev analytics is sent when the command is finished
- Add support to pass an `initialContext` to UI Task component. That way we can dynamically attach a method to the task implementation and pass the input arguments in the  `renderTasks` method

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx [hunk 1]
```diff
@@ -177,7 +178,7 @@ const ConcurrentOutput: FunctionComponent<ConcurrentOutputProps> = ({
           )
         }}
       </Static>
-      {footer ? (
+      {footer && !isAborted ? (
```
[the remark this answers is not in the document]
Comment (matteodepalo) [reply packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:181]: Good change, I think when I removed the check on the `state` I also removed this by mistake, but if someone aborts then it doesn't make sense to keep the footer.

File: packages/app/src/cli/services/dev.ts [hunk 2]
```diff
@@ -274,12 +276,20 @@ async function dev(options: DevOptions) {
...
-  await reportAnalyticsEvent({config: options.commandConfig})
-
+  const abortController = new AbortController()
+  abortController.signal.addEventListener('abort', async () => {
+    setInterval(() => {
+      treeKill('SIGINT')
+    }, GRACEFUL_EXIT_TIMEOUT)
+    outputInfo('\n\nInitiating graceful shutdown ...')
+    await disableDeveloperPreview({app: {...remoteApp, apiSecretKeys: []}, token, hiddenOutput: true})
+    tunnelClient?.stopTunnel()
```
[the remark this answers is not in the document]
Comment (amcaplan) [reply packages/app/src/cli/services/dev.ts:286]: It's not a promise. https://github.com/Shopify/cli/blob/abc4ce838bc1b69cf4eac37584e24053e92269ad/packages/plugin-cloudflare/src/tunnel.ts#L55-L57

But I think both will take some time, so we should await both. Maybe I need to wrap in a function for this to work, like:
```suggestion
    await Promise.all([
      disableDeveloperPreview({app: {...remoteApp, apiSecretKeys: []}, token, hiddenOutput: true}),
      (async () => { tunnelClient?.stopTunnel() })()
    ])
```

[the remark this answers is not in the document]
Comment (matteodepalo) [reply packages/app/src/cli/services/dev.ts:286]: If you don't await inside async I don't think it's going to change things in this case. If the function doesn't return a promise then putting it inside Promise.all shouldn't have any effect.

**Comment by amcaplan** [comment]
Going to close as we're leaning toward a different direction.

**Pull request closed without merging**